module/apmotel/span.go

```diff
@@ -0,0 +1,330 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//go:build go1.18
+// +build go1.18
+
+package apmotel // import "go.elastic.co/apm/module/apmotel/v2"
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"reflect"
+	"runtime"
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	semconv "go.opentelemetry.io/otel/semconv/v1.17.0"
+	"go.opentelemetry.io/otel/trace"
+
+	"go.elastic.co/apm/module/apmhttp/v2"
+	"go.elastic.co/apm/v2"
+)
+
+type event struct {
+	Name       string
+	Attributes []attribute.KeyValue
+	Time       time.Time
+}
+
+type status struct {
+	Code        codes.Code
+	Description string
+}
+
+type span struct {
+	mu sync.Mutex
+
+	provider *tracerProvider
+
+	startTime   time.Time
+	attributes  []attribute.KeyValue
+	events      []event
+	spanContext trace.SpanContext
+	status      status
+
+	tx   *apm.Transaction
+	span *apm.Span
+}
+
+func (s *span) End(options ...trace.SpanEndOption) {
```

Comment:
Should end hold a write lock? it may be called concurrently twice and may attempt to set the duration and other span / tx fields.

Also, if end is called concurrently with `SetAttributes` (which holds a lock), It may cause race conditions when ranging over the stored attributes and attempting to set them at the same time.